### PR TITLE
Add wp cli command to active PR branches

### PR DESCRIPTION
**Pull request by brbrr**
to use: `wp jetpack-beta branch activate branch_name` 

This feature will be used for creating Jurassic.ninja sites: https://github.com/oskosk/jurassic.ninja/pull/74

### CODE REVIEW

**Comment by enejb** [comment]
Can we do the following here? 
1. remove the PR part. 
So that the command is only `wp jetpack-beta branch activate branch_name`
2. provide a way to list all possible branch names that are currently available via new 
`wp jetpack-beta branch` 

If you want I can help with number 2 in a different PR. Also if we need to refactor the code so that it is more usable for the cli commands lets do that as well. :)

**Comment by oskosk** [comment]
Can we be more specific with the message shown when trying to activate a branch that is not on the manifest ?

**Comment by oskosk** [comment]
Apart from my last comment, it would be great if we could stop the user even before launching the site in the branch is not on the manifest instead of having them wait for the site creation, the plugin installation and the failed attempt through wp-cli .

**Comment by brbrr** [comment]
Thanks for your thoughts guys! I'll try to implement all that you mention, and if I'll run into trouble I know whom to ping ;)

> Can we be more specific with the message shown when trying to activate a branch that is not on the manifest ?
@oskosk Are you talking about `Error: Error Downloading: - Error: Invalid URL Provided.`? If yes - that would be problematic since `jetpack-beta` uses `wp_die()` which is unable to catch. We need to re-write that part to set a custom message. 

>  it would be great if we could stop the user even before launching the site
We could add one more subcommand - `check branch_name` for that.

**Comment by oskosk** [comment]
> We could add one more subcommand - check branch_name for that.

@brbrr don't worry, I was mostly talking about the implementation on the JN code. Checking there if the branch the user requests is available and don't even create the site if it's not...

**Comment by brbrr** [comment]
> I was mostly talking about the implementation on the JN code. Checking there if the branch the user requests is available and don't even create the site if it's not...

Yeah, I got your point! But I missed that at that point it would be no site or plugin installed. I'll check how its done here, and will hopefully copy it into JN codebase

**Comment by brbrr** [comment]
@enejb @oskosk 

Could you take a look at the recent changes? I've decided to use `branch list` instead just `branch` as its more explicit (IMO). 

If `branch list` thing is OK - I'll add same functionality to JurassicNinja

File: class-jetpackbetaclicommand.php [hunk 1]
```diff
@@ -10,39 +10,49 @@ class JetpackBetaCliCommand extends WP_CLI_Command {
...
-				if ( is_wp_error( $retvalue ) ) {
-					return WP_CLI::error( __( 'Error', 'jetpack' ) . $retvalue->get_error_message() );
+				return $this->install_jetpack( 'master', 'master' );
+			} elseif ( 'stable' === $args[1] ) {
+				return $this->install_jetpack( 'stable', 'stable' );
+			} else {
+				$branch_name = str_replace( '/', '_', $args[1] ); 
+				$url = Jetpack_Beta::get_install_url($branch_name, 'pr');
+				if ( $url === null) {
+					return WP_CLI::error( __( 'Invalid branch name. Try `wp jetpack-beta branch list` for list of available branches', 'jetpack' ) );
```
Comment (enejb) [inline class-jetpackbetaclicommand.php]: ❤️

File: class-jetpackbetaclicommand.php [hunk 2]
```diff
@@ -51,13 +61,21 @@ private function validation_checks($args) {
...
 
-			if ( empty( $args[1] ) ) {
-				return WP_CLI::error( __( 'Specify branch name', 'jetpack' ) );				
+			if ( 'activate' === $args[0] && empty( $args[1] ) ) {
+				return WP_CLI::error( __( 'Specify branch name. Try `wp jetpack-beta branch list` for list of available branches', 'jetpack' ) );				
+			}
+		}
+
+		private function install_jetpack($branch, $section) {
+			$retvalue = Jetpack_Beta::install_and_activate( $branch, $section  );
```
Comment (enejb) [inline class-jetpackbetaclicommand.php]: Can we change `$retvalue` to something else. I think `$result` reads better.

File: class-jetpackbetaclicommand.php [hunk 3]
```diff
@@ -10,39 +10,49 @@ class JetpackBetaCliCommand extends WP_CLI_Command {
...
 			if ( 'master' === $args[1] ) {
-				$retvalue = Jetpack_Beta::install_and_activate( 'master', 'master' );
-				if ( is_wp_error( $retvalue ) ) {
-					return WP_CLI::error( __( 'Error', 'jetpack' ) . $retvalue->get_error_message() );
+				return $this->install_jetpack( 'master', 'master' );
+			} elseif ( 'stable' === $args[1] ) {
+				return $this->install_jetpack( 'stable', 'stable' );
+			} else {
+				$branch_name = str_replace( '/', '_', $args[1] ); 
+				$url = Jetpack_Beta::get_install_url($branch_name, 'pr');
```
Comment (enejb) [inline class-jetpackbetaclicommand.php]: Some spacing is missing.

File: class-jetpackbetaclicommand.php [hunk 4]
```diff
@@ -10,39 +10,49 @@ class JetpackBetaCliCommand extends WP_CLI_Command {
...
 		 *
 		 */
 		public function branch( $args ) {
 
 			$this->validation_checks( $args );
+
+			if ( 'list' === $args[0] ) {
+				$manifest = Jetpack_Beta::get_beta_manifest();
+				$branches = ['stable', 'master', 'rc'];
+				foreach(get_object_vars($manifest->pr) as $key)
```
Comment (enejb) [inline class-jetpackbetaclicommand.php]: Spaceing missing here as well.

**Comment by enejb** [comment]
Found some minor issues. 
Will test the PR once they are fixed. 
❤️ the new error messages

**Comment by oskosk** [comment]
Is the PR description for the command up to date ?

**Comment by enejb** [comment]
Can we list each branch on a new line? 
Currently we get this. 
Also it would be nice to show the current active branch in that list by adding an asteric right nxt to it. 
Also listing them in alphabetical order would help. :) 

```Available branches: stable, master, rc, simple-payments/button-styles, issue-8162, update/do-not-signout-wpcom-by-default-masterbar, add-subject-filter, try/woocommerce-analytics, sync/fix-trash-post-one-event, JavaDevVictoria-patch-2, add/carousel-media-notice, social-icons-widget, social-icons-features, fix/yarn-build-production, remove/ismounted, patch-2, fix/8309, update/top-posts-images, add/comments/moderation-email-overrides, fix/widget-minor-conditions-dropdown-entities, add/jpo-save-business-address, add/changelog-58, update/sync-post-data, update/jpo-get-endpoint-non-flat, sync/blowery/r167670-wpcom-1515085964, update/lazy-images-add-ui, feature/active-modules-filter, validate_string_and_url, fix/sync-network-update-activity, remove/webpack-config-from-svn, update/search-post-type-selection, remove/create_function-usage, fix/widget-extend-and-fix-on-update, add/install-and-activate-plugin-function, fix/count-error-brightcove-attributes-php72, remove/_inc-root-child, update/yarn-docs, fix/paths-in-connection-banner-images, add/prompt-professional-for-jetpack-search, update/sync-users-enqueue-order, fix/plugins-callables, add/sync-user-password-changes, undefine-index-error```

Activation worked as expected. 
It would be nice to output something like. Some more output like. 
`Activating $branch...`
:) 

Feel free to make the changes requested above in a new PR. I don't want to hold this up :+1:

File: class-jetpackbetaclicommand.php [hunk 5]
```diff
@@ -51,13 +61,21 @@ private function validation_checks($args) {
 				return WP_CLI::error( __( 'Specify subcommand', 'jetpack' ) );
```
Comment (enejb) [inline class-jetpackbetaclicommand.php]: It would be nice to say what the subcommands are. Since I got this error and I wasn't sure what to do next.

**Comment by brbrr** [comment]
@oskosk description updated. thanks for pointing me out

@enejb Now output is looks like this:
```
# wp --allow-root jetpack-beta branch list
Available branches: 
JavaDevVictoria-patch-2
add-subject-filter
add/carousel-media-notice
add/changelog-58
add/comments/moderation-email-overrides
add/install-and-activate-plugin-function
add/jpo-save-business-address
add/prompt-professional-for-jetpack-search
add/sync-user-password-changes
feature/active-modules-filter
fix/8309
fix/count-error-brightcove-attributes-php72
fix/plugins-callables
fix/widget-extend-and-fix-on-update
fix/widget-minor-conditions-dropdown-entities
fix/yarn-build-production
issue-8162
master
patch-2
rc
remove/_inc-root-child
remove/create_function-usage
remove/ismounted
remove/webpack-config-from-svn
simple-payments/button-styles
social-icons-features
social-icons-widget
stable
sync/blowery/r167670-wpcom-1515085964
sync/fix-trash-post-one-event
try/woocommerce-analytics
undefine-index-error
update/do-not-signout-wpcom-by-default-masterbar
* update/jpo-contact-keep-state
update/jpo-enable-widgets
update/jpo-get-endpoint-non-flat
update/jpo-homepage-keep-state
update/lazy-images-add-ui
update/search-post-type-selection
update/sync-post-data
update/sync-users-enqueue-order
update/top-posts-images
update/yarn-docs
validate_string_and_url
```

**Comment by enejb** [comment]
That looks great. 
Can we update to make it look more like 
```
  sync/whitelist-user-capabilities
  update/do-full-sync-on-site-registration
  update/plugin-api-endpoint
  update/plugin-v1.2-update
  update/plugins-api-endpoints-v1.2
  update/plugins-api-v1.2-tweeks
* update/sync-post-data
  update/sync-post-types
  update/sync-protect-failed-login-attempts
  update/sync-user-login
  update/sync-users-enqueue-order
  updates/new-protect-from-styles
``` 
so indent the branches? Sorry for the nit picking :)

File: class-jetpackbetaclicommand.php [hunk 6]
```diff
@@ -10,53 +10,81 @@ class JetpackBetaCliCommand extends WP_CLI_Command {
 		 *
 		 * ## OPTIONS
 		 *
-		 *
 		 * activate master: Get a version of the master branch built every 15 minutes
-		 *
 		 * activate stable: Get the latest stable version of Jetpack
-		 *
+		 * activate branch_name: Get a version of PR. PR must be built and unit-tested before it become availabe
+		 * activate list: Get list of available jetpack branches to install
```
Comment (enejb) [inline class-jetpackbetaclicommand.php]: Should we change `activate list` to just `list`:

**Comment by enejb** [comment]
:shipit: Thanks for all the changes

**Pull request merged**